### ISSUE TITLE
mlirDataFormatters.py has bugs

### ISSUE THREAD

**Issue by onlytheworld** [llvm/llvm-project#69873]
I am trying to debug the following code using lldb with DataFormatter scripts `llvm/utils/lldbDataFormatters.py` and `mlir/utils/lldb- scripts/mlirDataFormatters.py`.

```
class MyPass : public ::mlir::OperationPass<mlir::ModuleOp> {
  ...
}
void MyPass::runOnOperation() {
  MLIRContext &context = getContext();
  ConversionTarget target(context);
  ...
}
```
I set the first instruction of `runOnOperation` to a breakpoint and run the program.
Then I tried typing `p target`, but it got stuck.
![image](https://github.com/llvm/llvm-project/assets/37205103/47b9c0fa-8f5a-440e-a8d4-ef48ebf66449)
When I try to remove the script `mlir/utils/lldb-scripts/mlirDataFormatters.py` everything works fine, so I think the problem is in the [mlirDataFormatters.py](https://github.com/llvm/llvm-project/blob/main/mlir/utils/lldb-scripts/mlirDataFormatters.py).

**Comment by onlytheworld** [comment]
This bug is caused by expanding the `target`'s field `legalOperations` before it is initialized.
![image](https://github.com/llvm/llvm-project/assets/37205103/725e8c9a-058a-4eb2-8ad5-b12c09e32eee)

**Comment by Michael137** [comment]
Seems like it's expected behaviour if you're printing it before it's been initialised?

**Comment by onlytheworld** [comment]
> Seems like it's expected behaviour if you're printing it before it's been initialised?

Yes, in fact I'm using the plugin codelldb for debugging on vscode, and at the breakpoints the plugin will try to print all variables, including uninitialized ones.

**Comment by Michael137** [comment]
```
(lldb) v
(std::map<int, int>) v = size=4294967295 {
  [0] = (first = 1, second = 1876946136)
  ...
}
*** Some of the displayed variables have more members than the debugger will show by default. To show all 
of them, you can either use the --show-all-children option to frame variable or raise the limit by changin
g the target.max-children-count setting.
```

For `std::map` e.g., we have a limit on the children we show. So printing something with `UINT_MAX` isn't an issue. We should do the same for the formatter. Not sure why it isn't triggering in this case. Possibly a setting we need to enable (?)

**Comment by Michael137** [comment]
@onlytheworld is it actually hanging indefinitely or does it eventually print the `255` elements and displays a warning? When I try printing an unitinialized `MapVector` it does the latter.

**Comment by onlytheworld** [comment]
It's hanging indefinitely. I can't pinpoint the cause of the problem, but I can give a way to trigger it. 
My lldb version is 17.0.0-custom provided by codelldb. Use the following command to debug

`lldb --local-lldbinit -- mlir-opt -lower-affine --split-input-file mlir/test/Conversion/AffineToStandard/lower-affine-to-vector.mlir`
My `.lldbinit` file is
```
command script import llvm/utils/lldbDataFormatters.py
command script import mlir/utils/lldb-scripts/mlirDataFormatters.py
```

The breakpoint is set in the first instruction of the function `LowerAffinePass::runOnOperation()` in the file `mlir/lib/Conversion/AffineToStandard/AffineToStandard.cpp`

Then the problem will be triggered.

**Comment by onlytheworld** [comment]
> @onlytheworld is it actually hanging indefinitely or does it eventually print the `255` elements and displays a warning? When I try printing an unitinialized `MapVector` it does the latter.

It's hanging indefinitely. I can't pinpoint the cause of the problem, but I can give a way to trigger it. 
My lldb version is 17.0.0-custom provided by codelldb. Use the following command to debug

`lldb --local-lldbinit -- mlir-opt -lower-affine --split-input-file mlir/test/Conversion/AffineToStandard/lower-affine-to-vector.mlir`

My `.lldbinit` file is
```
command script import llvm/utils/lldbDataFormatters.py
command script import mlir/utils/lldb-scripts/mlirDataFormatters.py
```

The breakpoint is set in the first instruction of the function `LowerAffinePass::runOnOperation()` in the file `mlir/lib/Conversion/AffineToStandard/AffineToStandard.cpp`

Then the problem will be triggered.

**Comment by Michael137** [comment]
Yup I can repro this on top-of-tree, thanks!

**Comment by onlytheworld** [comment]
@Michael137  I solved the problem by limiting num_buckets to a certain size, thank you very much for your help!

**Comment by onlytheworld** [comment]
> > @Michael137 I solved the problem by limiting num_buckets to a certain size, thank you very much for your help!
> 
> np! we'll want to find a general fix for it though

I found a similar issue in `mlirDataFormatters.py`.

```
class IPListRangeSynthProvider:
    """Define an LLDB synthetic children provider for an IPList."""

    def __init__(self, valobj, internal_dict):
        self.valobj = valobj
        self.update()

    def num_children(self):
        sentinel = self.valobj.GetChildMemberWithName("Sentinel")
        sentinel_addr = sentinel.AddressOf().GetValueAsUnsigned(0)

        # Iterate the next pointers looking for the sentinel.
        count = 0
        current = sentinel.GetChildMemberWithName("Next")
        while current.GetValueAsUnsigned(0) != sentinel_addr:
            current = current.GetChildMemberWithName("Next")
            count += 1

        return count
```
The while loop might not terminate if it's not initialized.
